001-clarifications-eval.txt

```diff
@@ -0,0 +1,754 @@
+
+        If it is the currently executing call frame, then it is termed
+        the top call frame.
+
+    4. A current program location
+
+        The target architecture program location corresponding to the
+        current call frame of the current thread.
+
+        The program location of the top call frame is the target
+        architecture program counter for the current thread. The call
+        frame information is used to obtain the value of the return
+        address register to determine the program location of the other
+        call frames (see 6.4 Call Frame Information).
+
+        It is required for the evaluation of location list expressions
+        to select amongst multiple program location ranges. It is
+        required for operations that specify target architecture
+        registers to support virtual unwinding of the call stack (see
+        6.4 Call Frame Information).
+
+        If specified:
+
+          * If the current call frame is the top call frame, it must be
+            the current target architecture program location.
+
+          * If the current call frame F is not the top call frame, it
+            must be the program location associated with the call site
+            in the current caller frame F that invoked the callee frame.
+
+          * Otherwise the result is undefined.
+
+    5. A current compilation unit
+
+        The compilation unit debug information entry that contains the
+        DWARF expression being evaluated.
+
+        It is required for operations that reference debug information
+        associated with the same compilation unit, including indicating
+        if such references use the 32-bit or 64-bit DWARF format. It can
+        also provide the default address space address size if no
+        current target architecture is specified.
+
+        [non-normative] For example: the DW_OP_constx and DW_OP_addrx
+        operations.
+
+        [non-normative] Note that this compilation unit may not be the
+        same as the compilation unit determined from the loaded code
+        object corresponding to the current program location. For
+        example, the evaluation of the expression E associated with a
+        DW_AT_location attribute of the debug information entry operand
+        of the DW_OP_call* operations is evaluated with the compilation
+        unit that contains E and not the one that contains the
+        DW_OP_call* operation expression.
+
+    6. A current target architecture
+
+        The target architecture.
+
+        It is required for operations that specify target architecture
+        specific entities.
+
+        [non-normative] For example: target architecture specific
+        entities include DWARF register identifiers, DWARF address space
+        identifiers, the default address space, and the address space
+        address sizes.
+
+        If specified:
+
+          * If the current thread is specified, then the current target
```

Comment:
On ARM are the modes considered different architectures, or just modes of a single architecture?

Another example where frames may have different architectures in the Sony Cell processor. I believe it could have host frames and cell frames on a single stack.

How would an implementation determine the architecture of the frame? Clearly the thread must know its architecture. Would it use the return address to determine the code object from which it could get an architecture? What if the code is dynamically generated and does not belong to a code object?

Reply:
What about changing the first bullet to:

          * If the current frame is specified, then the current target
            architecture must be the same as the target architecture of
            the current frame.

          * If the current frame is specified and is the top frame, and if
            the current thread is specified, then the current target
            architecture must be the same as the target architecture of
            the current thread.